### ISSUE TITLE
C&W fails to generate attacks on imagenet

### ISSUE THREAD

**Issue by itaim5** [jeromerony/fast_adversarial#9]
Hi,

First of all, I wanted to say that your paper is very interesting and enlightening, and I had a great time reading it. 

I used your C&W pytorch implementation on MNIST, FMNIST and CIFAR10 (with the default hyper-parameters) and generated attacks with 100% success rate.
However, when I tried to generate attacks for ImageNet, I failed every time and all my attempts were futile.
Do you have any idea why? 
it can be the hyper-parameters but I used various values for initial_const and LR, which are the only ones that seemed relevant.
I have to submit my paper in 1 week so this is urgent and I'll highly appreciate your help :)

Thanks,
Itai

**Comment by jeromerony** [comment]
Hi,

As mentioned in the README, this repo should not be used anymore. Instead, you should use the adversarial-library https://github.com/jeromerony/adversarial-library that is maintained properly for newer versions of PyTorch.

Without knowing more details about your experimental setup, I cannot help you more than that. If you do not wish to discuss details in public, you can send me a mail.

Cheers

**Comment by itaim5** [comment]
First of all, thank you very much for your reply.

I'm using your code in a very straight-forward way:
cwattacker = CarliniWagnerL2(device=device, image_constraints=(0, 1), num_classes=num_classes)
cw_attack = cwattacker.attack(model, x_image, labels=torch.tensor([target_label.item()], device=device), targeted=True)
I was able to generate successful attacks at a 100% success rate for MNIST, Fashion-MNIST and CIFAR-10, by using it the standard way (without even changing any of the default parameters).  
For ImageNet, the only change I had to do relates to the normalization that has to be done (I'm using a pre-trained resnext-50 model that expects normalization), and because of that I wrapped the forward call inside your code (in the step and quantize functions. although I'm not using quantization) with a normalization operation:
x_image = (x_image - mean) / std
I don't think the normalization is the issue here (as I used it successfully the same way when generating my own types of attacks), but it's the only thing I did differently for ImageNet, so I had to state that.

So, for some reason I'm not able to produce even one successful attack for ImageNet using this method. I tried playing with different values of CarliniWagnerL2's parameters (mostly lr and initial_const) and it didn't help.
Initially I thought that something is off and that maybe the gradients zero out, but it's not the case as I see that the modifier keeps changing along the optimization process.

BTW - I'm using it for a single attack at a time (just to clarify why you see a scalar tensor in labels) - a batch of 1.

Any ideas?

**Comment by jeromerony** [comment]
The normalization is most likely the culprit here. Try using the `NormalizedModel` wrapper from `fast_adv.utils` instead of modifying the code of the C&W attack.

**Comment by itaim5** [comment]
Thanks, now I see your implementation of a normalized model wrapper.
I did implement a forward pass wrapper for my model which does exactly what you do (without using nn.Parameter).
I'll try that, but why do you think this is the culprit?

**Comment by itaim5** [comment]
just for future reference - it wasn't the normalization, so I have no idea why it failed. 
I ended up using another C&W implementation.
